### PR TITLE
Fix workspace color picker context menu regression

### LINKED ISSUE #4646: Workspace color picker dismisses before color can be selected (regression of #2560)
## Bug

Most of the time when right-clicking a workspace and hovering "Color", the color submenu disappears before a color can be selected. The user has to retry several times before the menu stays open long enough to click a swatch.

This matches the symptom of [#2560](https://github.com/manaflow-ai/cmux/issues/2560) ("Workspace color picker menu blinks in and out, impossible to select a color"), which was closed by [#2566](https://github.com/manaflow-ai/cmux/pull/2566) on 2026-04-07. The fix in #2566 deferred `TabItemView`'s `workspaceObservationGeneration` increments while the context menu was visible, so debounced workspace updates wouldn't rebuild `.contextMenu` and dismiss the submenu mid-interaction.

The user is hitting it again now — same symptom, with the fix in place. So either:

1. **The deferral doesn't catch every update path.** #2566 routed updates through `scheduleWorkspaceObservationInvalidation()` and short-circuited on `contextMenuVisible`. Anything else that mutates `TabItemView`'s body — a state change outside that helper, a parent re-render, an `@Published` from a different observable — will still rebuild `.contextMenu` and dismiss the submenu. Worth auditing every property the row reads.
2. **`.contextMenu`'s `.onAppear` is not firing reliably on macOS 26.x.** The fix relies on `.onAppear` inside the `.contextMenu { }` content to flip `contextMenuVisible = true`. If macOS 26's submenu lifecycle skips that, `contextMenuVisible` stays `false`, no deferral happens, and the original blinking returns. The fix's flowchart already calls out the inverse case (onDisappear skipped → stuck true); the onAppear-skipped case would manifest exactly as this bug.
3. **Cross-process publisher (Convex / WebSocket) fires while the menu is up.** If sidebar state pulls from a Combine publisher that re-emits independent of the deferred helper, the deferral never sees it.

Hard to reproduce reliably — it happens "most of the time" but not always, which suggests timing/race rather than a deterministic state bug. Likely needs `$cloud-mac` to record video on a clean VM that matches my macOS 26.4.1 / cmux 0.64.9 setup.

## Steps to reproduce

1. Right-click a workspace in the sidebar.
2. Hover or click "Color" to open the color submenu.
3. Try to click a color swatch.
4. **Expected:** menu stays open until selection or click-away.
5. **Actual:** submenu dismisses (often immediately, before a click registers); user retries 3–5 times before it sticks.

## Environment

- macOS 26.4.1 (build 25E253) — i.e. macOS Tahoe point release
- cmux 0.64.9 (build 89), `/Applications/cmux.app`
- No DEV/RC build — stable channel

## Suspect files

- `Sources/ContentView.swift` — `TabItemView` `scheduleWorkspaceObservationInvalidation()` / `contextMenuVisible` (added in #2566)
- Anywhere `TabItemView` reads from a publisher not routed through that helper.

## Related

- #2560 (closed, same symptom)
- #2566 (fix, merged 2026-04-07)
- #3081 (different "Left Rail" color regression)

cc: should reference #2560 as a probable regression of the same root cause class.

### PR DESCRIPTION
**Pull request by austinywang**
Fixes #4646.

Regression context: #2560 was fixed by #2566 on 2026-04-07 by freezing `TabItemView` context-menu presentation and deferring workspace observation updates while the menu is open.

## Summary
- Restore full frozen sidebar row presentation while a workspace context menu is active so unread badge/preview and shortcut-hint publishes cannot change `TabItemView` equality and rebuild `.contextMenu` mid-submenu.
- Replace the menu visibility boolean with a single row menu lifecycle that treats AppKit `NSMenu` tracking and SwiftUI `.contextMenu` content as the same active phase. This covers the macOS 26 case where SwiftUI `.onAppear` may be late or skipped, and the inverse `.onDisappear`-skipped case.
- Start and stop the presentation freeze from the AppKit tracking callback too, so the freeze begins before SwiftUI menu content lifecycle callbacks are required.

## Root Cause
#4084 made frozen sidebar presentations keep unread count and latest notification text live during a context-menu freeze. Those fields participate in `TabItemView ==`, so notification-store publishes could still re-evaluate the row and rebuild `.contextMenu` while the Workspace Color submenu was open. Separately, the deferral gate only trusted SwiftUI `.contextMenu` `.onAppear`, while the row already had a more reliable AppKit menu-tracking signal.

## Regression Coverage
Two-commit red/green structure:
1. `e4f809c0d` changes/adds failing runtime policy tests for frozen presentation stability and AppKit menu-tracking visibility.
2. `515b8e45c` implements the fix.

## Verification
- `git diff --check`

Not run locally per repo/task policy: local tests/builds and `./scripts/reload.sh`.

Cloud Mac video: unavailable. `cloud-mac preflight` is currently blocked locally by unauthenticated `macfleet`, and `gh auth status` reports invalid keyring tokens. I still pushed the branch successfully over the repo remote, but cloud VM provisioning could not be started from this checkout.

---
<a href="https://app.blacksmith.sh/manaflow-ai/codesmith/cmux/pr/4648"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1782117481&installation_id=133855650&pr_number=4648&repository=manaflow-ai%2Fcmux&return_to=https%3A%2F%2Fgithub.com%2Fmanaflow-ai%2Fcmux%2Fpull%2F4648&signature=81e4321ec85f19161c8d3b17aa82a8293ed69e8a1c78e5868345b24d4031e38d"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need. Autofix is disabled.</sup>

---

> [!NOTE]
> **Medium Risk**
> Touching sidebar row/menu interaction state and snapshot freezing logic could introduce subtle UI regressions (hover/close button visibility or deferred snapshot flushing), though changes are localized and backed by new tests.
> 
> **Overview**
> Fixes a sidebar workspace context-menu regression by **fully freezing** `SidebarTabItemPresentationSnapshot` while a row menu is active, preventing live unread/notification updates from rebuilding `TabItemView` (and its `.contextMenu`) mid-interaction.
> 
> Unifies menu visibility handling by replacing the simple boolean with a **SwiftUI + AppKit context-menu lifecycle** in `SidebarWorkspaceRowInteractionState`, and starts/ends the freeze based on both SwiftUI appear/disappear and AppKit `NSMenu` tracking callbacks (via a new `onContextMenuTrackingChanged` hook in `SidebarWorkspaceRowHoverTracker`). Adds/updates tests to cover the new freeze semantics and menu lifecycle edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6de0bbea9fd5317418475b52331b3426a02b3fd6. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
## Summary by cubic
Fixes the workspace color picker context menu regression by fully freezing the sidebar row while any menu is active and deriving visibility from a unified SwiftUI/AppKit lifecycle. Stops unread/preview changes from rebuilding the menu mid-submenu and keeps hover suppression until AppKit tracking ends.

- **Bug Fixes**
  - Return the frozen `SidebarTabItemPresentationSnapshot` during an active menu to stabilize unread count, preview text, and shortcut hints.
  - Start/end the freeze from AppKit `NSMenu` tracking and SwiftUI; ignore SwiftUI disappear while tracking and flush deferred workspace updates only after tracking ends; `SidebarWorkspaceRowHoverTracker` now reports tracking changes to drive the freeze.

- **Refactors**
  - Replace the boolean menu flag with a lifecycle state (inactive, SwiftUI-visible, AppKit-tracking) in `SidebarWorkspaceRowInteractionState` to unify visibility and close-button suppression across edge cases.

<sup>Written for commit 6de0bbea9fd5317418475b52331b3426a02b3fd6. Summary will update on new commits. <a href="https://cubic.dev/pr/manaflow-ai/cmux/pull/4648?utm_source=github">Review in cubic</a></sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Sidebar badges and notification text remain stable while context menus are open.
  * Context-menu interactions now preserve hover/close-button behavior and visibility across different menu lifecycles.
  * Sidebar presentation now consistently maintains state during menu operations.

* **Tests**
  * Added tests covering context-menu visibility and notification-state stabilization across interaction patterns.

[Review Change Stack](https://app.coderabbit.ai/change-stack/manaflow-ai/cmux/pull/4648?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: Sources/Sidebar/SidebarWorkspaceSnapshotRefreshPolicy.swift [hunk 1]
```diff
@@ -94,25 +104,31 @@ struct SidebarWorkspaceRowInteractionState: Equatable {
...
     }
 
     mutating func contextMenuDidDisappear() {
-        contextMenuVisible = false
+        if contextMenuLifecycle != .appKitTracking {
+            contextMenuLifecycle = .inactive
+        }
         contextMenuTrackingSuppressesCloseButton = false
         applyDeferredPointerHovering()
     }
```
[the remark this answers is not in the document]
Comment (austinywang) [reply Sources/Sidebar/SidebarWorkspaceSnapshotRefreshPolicy.swift:120]: Fixed in 362343cc3: contextMenuDidDisappear now returns while AppKit menu tracking is authoritative, so hover suppression is released only by contextMenuTrackingDidEnd.

*— Claude Code*

[the remark this answers is not in the document]
Comment (austinywang) [reply Sources/Sidebar/SidebarWorkspaceSnapshotRefreshPolicy.swift:120]: Fixed in 362343cc3 with the same lifecycle guard, plus regression coverage for SwiftUI disappearance during active AppKit tracking.

*— Claude Code*

File: Sources/ContentView.swift [hunk 2]
```diff
@@ -13263,6 +13258,7 @@ struct SidebarWorkspaceSnapshotBuilder {
 }
 
 private final class SidebarTabItemContextMenuState: ObservableObject {
+    var presentationFreezeActive = false
     var hasDeferredWorkspaceObservationInvalidation = false
     var pendingWorkspaceSnapshot: SidebarWorkspaceSnapshotBuilder.Snapshot?
 }
```
[the remark this answers is not in the document]
Comment (austinywang) [reply Sources/ContentView.swift]: Fixed in 362343cc3 by marking SidebarTabItemContextMenuState as @MainActor; its SwiftUI/AppKit menu lifecycle accesses are main-actor UI state.

*— Claude Code*

[the remark this answers is not in the document]
Comment (austinywang) [reply Sources/ContentView.swift]: Updated in 6de0bbea9: instead of keeping this as an @MainActor ObservableObject, I removed the object entirely and made the non-published context-menu storage a plain @State value. The freeze-active bit is now derived from the frozen presentation binding.

*— Claude Code*
